Remove need for upsert_all

kaspth opened the pull request:
In light of #35636 we're trying to extend upsert/upsert_all to handle both an array of columns or some update SQL.

This tries to tackle this problem in a different by replacing upsert_all with an upsert option on insert_all:

```ruby
Book.insert_all rows, upsert: true
Book.insert_all rows, upsert: %i( id name author_id )
Book.insert_all rows, upsert: "status = GREATEST(books.status, excluded.status)"
```

The latter two are not added in this PR, but that's how they'd look.

upsert_all still looks cleaner for the plain true case, so perhaps this would be better:

```ruby
Book.upsert_all rows, only: %i( id name author_id )
Book.upsert_all rows, with: "status = GREATEST(books.status, excluded.status)"
```

Though the minute differences between insert_all/insert_all! and upsert_all are quite hard to distinguish. And perhaps a method fewer is worth it for what is a rare call.

cc @boblail @palkan @janko @abhaynikam

kaspth committed:
Remove the need for upsert/upsert_all

In light of https://github.com/rails/rails/pull/35636 we're trying to
extend upsert/upsert_all to handle both an array of columns or some
update SQL.

This tries to tackle this problem in a different by replacing upsert_all
with an upsert option on insert_all:

```ruby
Book.insert_all rows, upsert: true
Book.insert_all rows, upsert: %i( id name author_id )
Book.insert_all rows, upsert: "status = GREATEST(books.status, excluded.status)"
```

The latter two are not added in this PR, but that's how they'd look.

upsert_all still looks cleaner for the plain true case, so perhaps this
would be better:

```ruby
Book.upsert_all rows, only: %( id name author_id )
Book.upsert_all rows, with: "status = GREATEST(books.status, excluded.status)"
```

Though the minute differences between insert_all/insert_all! and
upsert_all are quite hard to distinguish. And perhaps a method fewer is
worth it for what is a rare call.

kaspth committed:
Mark InsertAll as private API. Easier to add later.

kaspth committed:
Capture some join calls.

kaspth commented on activerecord/lib/active_record/insert_all.rb:
```diff
@@ -73,15 +75,11 @@ def ensure_valid_options_for_connection!
           raise ArgumentError, "#{connection.class} does not support :returning"
         end
 
-        unless %i{ raise skip update }.member?(on_duplicate)
```
We own the callers. So no need for this.

kaspth commented on activerecord/lib/active_record/insert_all.rb:
```diff
@@ -137,28 +135,31 @@ def values_list
+        def upsert_sql(&block)
+          case insert_all.upsert
+          when true
+            quote_columns(insert_all.updatable_columns).map(&block).join(",")
```
@palkan https://github.com/rails/rails/pull/35636 would essentially be cut down to just extending this with:

```ruby
when Array
  quote_columns(insert_all.upsert).map(&block).join(",")
when String
  insert_all.upsert
end
```

kaspth commented on activerecord/test/cases/insert_all_test.rb:
```diff
@@ -11,6 +11,18 @@ class InsertAllTest < ActiveRecord::TestCase
+
+    Book.insert({ id: id, name: "Remote", author_id: 1 }, { upsert: true })
+
+    assert_equal "Remote", Book.find(id).name
```
We were missing tests for just plain `insert`.

kaspth commented:
On the one hand I really want to be done with `insert_all`. But on the other hand I still find it hard to explain to myself what all the different options mean and how they vary method to method. I'd rather not ship that and get confusing bug reports after release. Or maybe everything is just clear as day and I just don't get it! 😄

simi reviewed (approved):
:+1:

palkan commented on activerecord/lib/active_record/insert_all.rb:
```diff
@@ -34,11 +36,11 @@ def primary_keys
 
     def skip_duplicates?
-      on_duplicate == :skip
+      !@raise && !@upsert
```
Doesn’t it add `ON CONFLICT DO NOTHING` (PostgreSQL) even we haven’t specified any options? Looks so.

We don’t need to add _on conflict_ clause in that case.

Shouldn’t it be:

```suggestion
      !@raise && (@upsert != false)
```

kaspth replied:
You're thinking of `insert_all!`, `insert_all` (no-bang) skips by default.

palkan reviewed (commented):
I like the idea of having only `insert_all`.

Foe me it looks more clear when we have `insert_all(..., upsert: ...)` then having both `insert_all` / `upsert_all` (maybe, because I’ve been using the same API with my custom batch insert method, except from the keyword name–I use  `on_conflict` name).

And it’s probably better represent the underlying SQL: it is still `INSERT ...` with an additional scope; there is no such thing as `UPSERT ...`.

kaspth committed:
Fix SQLite3 test on Travis

It seems the Travis version of SQLite3 isn't >= 3.24.0 per
`supports_insert_on_conflict?`, so we'll need this skip.

kaspth commented:
The tough part about the current API as it's trying to match AR's naming conventions is that:

- `insert_all` skips duplicates by default, but can still raise if passed a `:unique_by`, matching "ON CONFLICT DO NOTHING".
- `insert_all!` always raises on duplicates, though has no support for `:unique_by`, ultimately matching no "ON CONFLICT" clause.
- `upsert_all` overwrites the values in duplicates, matching "ON CONFLICT UPDATE".

I'm coming back around to keeping `upsert_all` so these 3 strategies are nicely mapped out.

Also: we'd need to move `:unique_by` to `insert_all!` entirely.

boblail commented:
@kaspth, I agree, `insert_all` v. `insert_all!` is tricky since:

- `insert_all` could still raise (depending on the value of `unique_by`)
- the method with the bang (`insert_all!`) is the "natural" behavior (no ON CONFLICT clause)

OTOH:
- Having one method with a strategy arg was a scenario David steered me away from.
- And `upsert_all` is a pleasantly clear method name (and it reads very well in your examples above with `:only` and `:with` as the kwargs 👍)

* * *

🤔 Part of David's thinking was that bang-v-no-bang would parallel `save`/`save!`, `create`/`create!`, and other pairs in ActiveRecord.

But `insert_all` more directly parallels `update_all` and `delete_all` (which don't have bang counterparts)...

I wonder if — instead of choosing between 3 methods and 1 — there's a good case for 2 methods:

1.  `upsert_all` as you described above, Kasper, and
2. `insert_all` like this:

    ```ruby
    insert_all books
    insert_all books, skip_duplicates: true
    insert_all books, skip_duplicates: true, unique_by: :isbn
    ```
    or
    ```ruby
    insert_all books, raise_on_duplicates: false
    ```

    
    If we did that, though, would would _this_ example mean?
    
    ```ruby
    insert_all books, unique_by: :isbn
    ```
    
    Would `unique_by` be _imply_ `skip_duplicates` or would it be ignored if `skip_duplicates == false`?

palkan commented:
> But insert_all more directly parallels update_all and delete_all (which don't have bang counterparts)...

+1

`insert_all!` seems confusing, and it has a different meaning than `save!` / `create!`–it relies on DB-level exception, not framework level.

And having both `insert_all` / `insert_all!` could be confusing for users: which one should I choose and why? (And we will likely end up with a Rubocop cop `Rails/InsertAllBangAlways`).

@boblail Did you mean `insert_all` in 1) and `upsert_all` in 2) in the example above?

dhh commented:
Here's what I'd like to see:

1. insert_all never raises. Any configuration option is simply to instruct it in ways it'll ignore duplicates of various kinds.
1. insert_all! does all the raising. By default on duplicates, and by configuration in other cases.
1. upsert_all does the overwriting, but shouldn't need to raise. If we need a version of upsert_all that still DOES raise, we should consider adding upsert_all! to deal with that.

palkan commented:
> insert_all never raises. 

That could be surprising to the end user for the following reasons:
- first, as mentioned above, we do have `update_all` which could fail at DB-level (e.g. due to uniqueness constraints) – that's an expected behaviour for all Rails developers 
- secondly, even while using ORMs most developers "think" in queries, and thus `insert_all` naturally translates into `INSERT INTO ... VALUES ...`, but not into  `INSERT INTO ... VALUES ... ON CONFLICT DO NOTHING`.

So, I will have to keep in mind that `insert_all` is not just `INSERT`.

dhh commented:
To me, that's the point of difference between insert_all and insert_all!.
Do you want to fire and forget, or do you want to deal with exceptions? Of
course, if there's an exception related to something other than the
mechanics of the insert, we'll raise (like table is missing), but otherwise
not.

On Thu, Mar 21, 2019 at 8:45 AM Vladimir Dementyev <notifications@github.com>
wrote:

> insert_all never raises.
>
> That could be surprising to the end user for the following reasons:
>
>    - first, as mentioned above, we do have update_all which could fail at
>    DB-level (e.g. due to uniqueness constraints) – that's an expected
>    behaviour for all Rails developers
>    - secondly, even while using ORMs most developers "think" in queries,
>    and thus insert_all naturally translates into INSERT INTO ... VALUES
>    ..., but not into INSERT INTO ... VALUES ... ON CONFLICT DO NOTHING.
>
> So, I will have to keep in mind that insert_all is not just INSERT.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/35686#issuecomment-475284415>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtQ-kxE_ugIY-IuNjzJEOQQAQQpdrks5vY6kugaJpZM4cAHaF>
> .
>

palkan commented:
> Do you want to fire and forget, or do you want to deal with exceptions?

TBH, I can't see why `insert_all` implies "fire and forget".

Users **must** deal with potential race conditions, explicitly, we should not hide potential caveats.

That could affect the application logic. Here is a very simplified example just for demonstration:

```ruby
def invite_new_users(attrs)
  # attrs => [{ name: "Alice", email: "alice@rails.six" }, ...]
  User.insert_all(attrs)
  attrs.each { |data| UsersMailer.send_invitation(data.fetch(:email)) }
end
``` 

You know what's gonna happen if the author of this code has no idea about implicit _skip duplicates_ behaviour.

And I can't see how the name of the method–`insert_all`–tells me about this caveat (and the presence of the banged version is even more confusing–it might be considered as validations-aware version, 'cause we use `create!` / `save!` / `update!` to raise when validations fail).

If I want "fire and forget", I want it to be explicit, e.g. `insert_all(..., upsert: :skip_duplicates)`–that looks totally clear to me.

`UPSERT` itself is an advance feature, I don't think we should worry about adding additional parameter to make it work. We're no sacrificing usability here.

dhh commented:
I don't agree with this assessment of the usability. If you call
Model#save, you also have to know that validation errors will prevent that
save from happening, that the return value will merely be a false, and you
have to inspect the #errors yourself. If you want an exception raised, you
use #save!

This design follows the same philosophy.

On Thu, Mar 21, 2019 at 3:12 PM Vladimir Dementyev <notifications@github.com>
wrote:

> Do you want to fire and forget, or do you want to deal with exceptions?
>
> TBH, I can't see why insert_all implies "fire and forget".
>
> Users *must* deal with potential race conditions, explicitly, we should
> not hide potential caveats.
>
> That could affect the application logic. Here is a very simplified example
> just for demonstration:
>
> def invite_new_users(attrs)
>
>   # attrs => [{ name: "Alice", email: "alice@rails.six" }, ...]
>
>   User.insert_all(attrs)
>
>   attrs.each { |data| UsersMailer.send_invitation(data.fetch(:email)) }
> end
>
> You know what's gonna happen if the author of this code has no idea about
> implicit *skip duplicates* behaviour.
>
> And I can't see how the name of the method–insert_all–tells me about this
> caveat (and the presence of the banged version is even more confusing–it
> might be considered as validations-aware version, 'cause we use create! /
> save! / update! to raise when validations fail).
>
> If I want "fire and forget", I want it to be explicit, e.g. insert_all(...,
> upsert: :skip_duplicates)–that looks totally clear to me.
>
> UPSERT itself is an advance feature, I don't think we should worry about
> adding additional parameter to make it work. We're no sacrificing usability
> here.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/35686#issuecomment-475422641>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKtVggs1TEutcGfIU3WE31wv9Zt-94ks5vZAPKgaJpZM4cAHaF>
> .
>

palkan commented:
> This design follows the same philosophy.

Why it's not following the same philosophy as `update_all` then?)

`Model#save` is from the different level of abstraction than `Model.insert_all`.
If we do `insert_all` with all the model-related stuff like validations and callbacks (e.g. like a popular `activerecord-import` gem does), then it would make sense to follow the same _philosophy_.

That's just a helper method to build a mass insert query, it doesn't know anything about the _model_, it relies only on the DB metadata.

dhh commented:
I don’t think we are going to end up with a meeting of the mind here. I’ve laid out my thought process and made my assessment. Thank you for your input and your arguments 🙏

> On Mar 22, 2019, at 17:00, Vladimir Dementyev <notifications@github.com> wrote:
> 
> This design follows the same philosophy.
> 
> Why it's not following the same philosophy as update_all then?)
> 
> Model#save is from the different level of abstraction than Model.insert_all.
> If we do insert_all with all the model-related stuff like validations and callbacks (e.g. like a popular activerecord-import gem does), then it would make sense to follow the same philosophy.
> 
> That's just a helper method to build a mass insert query, it doesn't know anything about the model, it relies only on the DB metadata.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

simi commented:
I think `insert_all` (and friends) should not be compared to `save` + `errors` combo since they both operate on different level.

In my :eyes: `insert_all` is not primary meant to be used in controller facing user input as replacement for `create`/`find_or_initialize_by`+`save`. There is clear trade-off for `insert_all` methods -> you get performance benefit but you lose data validation and callbacks.

Similar to `update_all` method operating on same level, even `non-bang` methods don't work in "fire and forget" mode and they can result into various database related exceptions.

```
User.update_all(id: nil)
  SQL (0.7ms)  UPDATE "users" SET "id" = NULL
# => ActiveRecord::StatementInvalid (PG::NotNullViolation: ERROR:  null value in column "id" violates not-null constraint)
```

There's no easy way to ensure those database-level methods will never raise. I think current API is good trade-off to not have methods with a bunch of arguments and still keep it clear and descriptive. 

I suggest to accept fact those methods are more advanced to use. Don't confuse beginners in general rails guides with those and make them easily searchable in some advanced ActiveRecord guides describing all the caveats, links to supported adapters documentations and showing some examples.

kaspth commented:
Yeah, let's keep `upsert_all` then! Thanks everyone.
